Prevent building cargo from invalidating build cache of other tools due to conditionally applied `-Zon-broken-pipe=kill` via tracked `RUSTFLAGS`

jieyouxu opened the pull request:
This PR fixes #130980 where building cargo invalidated the tool build caches of other tools (such as rustdoc) because `-Zon-broken-pipe=kill` was conditionally passed via `RUSTFLAGS` for other tools *except* for cargo. The differing `RUSTFLAGS` triggered tool build cache invalidation as `RUSTFLAGS` is a tracked env var -- any changes in `RUSTFLAGS` requires a rebuild.

`-Zon-broken-pipe=kill` is load-bearing for rustc and rustdoc to not ICE on broken pipes due to usages of raw std `println!` that panics without the flag being set, which manifests in ICEs.

I can't say I like the changes here, but it is what it is...

See detailed discussions and history of `-Zon-broken-pipe=kill` usage in https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Internal.20lint.20for.20raw.20.60print!.60.20and.20.60println!.60.3F/near/474593815.

## Approach

This PR fixes the tool build cache invalidation by informing the `rustc` binary shim when to apply `-Zon-broken-pipe=kill` (i.e. when the rustc binary shim is not used to build cargo). This information is not communicated by `RUSTFLAGS`, which is an env var tracked by cargo, and instead uses an untracked env var `UNTRACKED_BROKEN_PIPE_FLAG` so we won't trigger tool build cache invalidation. We preserve bootstrap's behavior of not setting that flag for cargo by conditionally omitting setting `UNTRACKED_BROKEN_PIPE_FLAG` when building cargo.

Notably, the `-Zon-broken-pipe=kill` instance in https://github.com/rust-lang/rust/blob/1e5719bdc40bb553089ce83525f07dfe0b2e71e9/src/bootstrap/src/core/build_steps/compile.rs#L1058 is not modified because that is used to build rustc only and not cargo itself.

Thanks to @cuviper for the idea!

## Testing

### Integration testing

This PR introduces a run-make test for rustc and rustdoc that checks that when they do not ICE/panic when they encounter a broken pipe of the stdout stream.

I checked this test will catch the broken pipe ICE regression for rustc on Linux (at least) by commenting out https://github.com/rust-lang/rust/blob/1e5719bdc40bb553089ce83525f07dfe0b2e71e9/src/bootstrap/src/core/build_steps/compile.rs#L1058, and the test failed because rustc ICE'd.

### Manual testing

I have manually tried:

1. `./x clean && `./x test build --stage 1` -> `rustc +stage1 --print=sysroot | false`: no ICE.
2. `./x clean` -> `./x test run-make` twice: no stage 1 cargo rebuilds.
3. `./x clean` -> `./x build rustdoc` -> `rustdoc +stage1 --version | false`: no panics.
4. `./x test src/tools/cargo`: tests pass, notably `build::close_output` and `cargo_command::closed_output_ok` do not fail which would fail if cargo was built with `-Zon-broken-pipe=kill`.

## Related discussions

Thanks to everyone who helped!
- https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Applying.20.60-Zon-broken-pipe.3Dkill.60.20flags.20in.20bootstrap.3F
- https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/Modifying.20run-make.20tests.20unnecessarily.20rebuild.20stage.201.20.2E.2E.2E
- https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Internal.20lint.20for.20raw.20.60print!.60.20and.20.60println!.60.3F

Fixes https://github.com/rust-lang/rust/issues/130980
Closes https://github.com/rust-lang/rust/issues/131059

---

try-job: aarch64-apple
try-job: x86_64-msvc
try-job: x86_64-mingw

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11145306391/job/30974618835) [(plain)](https://github.com/rust-lang/rust/commit/e103cd0da04393c1ecdd2a9d870c71c92694d9cc/checks/30974618835/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#11 [ 4/12] RUN sh /scripts/nodejs.sh /node
#11 0.039 + NODEJS_VERSION=v20.12.2
#11 0.039 + INSTALL_PATH=/node
#11 0.039 + url=https://nodejs.org/dist/v20.12.2/node-v20.12.2-linux-x64.tar.xz
#11 0.039 + curl -sL+ tar https://nodejs.org/dist/v20.12.2/node-v20.12.2-linux-x64.tar.xz -xJ
#11 2.059 + mv node-v20.12.2-linux-x64 /node
#11 DONE 2.2s

#12 [ 5/12] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
---
#16 2.812 Building wheels for collected packages: reuse
#16 2.814   Building wheel for reuse (pyproject.toml): started
#16 3.059   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.060   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.060   Stored in directory: /tmp/pip-ephem-wheel-cache-yig0s9sg/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.062 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.456 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.457 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
##[group]Linting tool bootstrap (stage0 -> stage1, x86_64-unknown-linux-gnu)
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/clippy-driver /checkout/obj/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names --cfg=bootstrap --sysroot /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=cfg(bootstrap)' '--check-cfg=cfg(llvm_enzyme)' '--check-cfg=cfg(parallel_compiler)' '--check-cfg=cfg(rust_analyzer)' -Zforce-unstable-if-unmarked -Zmacro-backtrace -Csplit-debuginfo=off --cfg=parallel_compiler -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Zunstable-options --target x86_64-unknown-linux-gnu --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
  --- stderr
  error: multiple input filenames provided (first two filenames are `/checkout/obj/build/bootstrap/debug/rustc` and `-`)
Build completed unsuccessfully in 0:01:35
  local time: Wed Oct  2 14:16:16 UTC 2024
  network time: Wed, 02 Oct 2024 14:16:16 GMT
##[error]Process completed with exit code 1.
```

</details>

jieyouxu committed:
Use untracked env var to pass `-Zon-broken-pipe=kill` for tools

- Don't touch rustc's `-Zon-broken-pipe=kill` env var in `compile.rs`.
- Use an untracked env var to pass `-Zon-broken-pipe=kill` for tools but
  skip cargo still, because cargo wants `-Zon-broken-pipe=kill` unset.

jieyouxu committed:
Add regression test for rustc/rustdoc broken pipe ICEs

jieyouxu commented on tests/run-make/broken-pipe-no-ice/rmake.rs:
Remark: in theory we might also check `cargo`, but cargo's test suite does catch if `cargo` was built with `-Zon-broken-pipe=kill`.

jieyouxu commented:
r? @onur-ozkan since you said you'd like to review this. Thanks for the help as well!

Noratrieb commented on tests/run-make/broken-pipe-no-ice/rmake.rs:
```diff
@@ -0,0 +1,51 @@
+
+fn main() {
+    let mut rustc = Command::new(env_var("RUSTC"));
+    rustc.arg("--print=sysroot");
```
but if `-Zon-broken-pipe=kill` is unset, `safe_println` will still emit an IO error, right? which is wrong

jieyouxu replied:
> but if `-Zon-broken-pipe=kill` is unset, `safe_println` will still emit an IO error, right? which is wrong

Yeah idk how specifically that will be handled, but that is left as an exercise for the future reader.

jieyouxu commented:
cc @rust-lang/cargo since this PR should not affect how cargo is built (i.e. cargo should still not get `-Zon-broken-pipe=kill`, would appreciate a double-check.

cc @rust-lang/rustdoc because rustdoc is also built (both before this PR and after) with `-Zon-broken-pipe=kill`, and the regression test exercises the rustdoc binary and checks that it does not panic upon broken pipe. It is not clear to me if this is the desired behavior for rustdoc, hence this FYI.

ehuss commented on tests/run-make/broken-pipe-no-ice/rmake.rs:
```diff
@@ -0,0 +1,50 @@
+use run_make_support::{env_var, libc};
+
+fn check_broken_pipe_handled_gracefully(name: &str, mut cmd: Command) {
+    cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
+
+    let mut child = cmd.spawn().unwrap();
+    // Close stdout
+    drop(child.stdout.take());
```
I would recommend caution with this approach, since it is inherently racey. For example, it assumes that it can close the handle before `rustc` is able to print. There can also be assumptions here about `PIPE_BUF` and other pipe behaviors. This can be unreliable on some systems. For example, on macOS, this should fail occasionally. 

That's why cargo's tests which exercise this same behavior use a proc-macro to synchronize with `rustc`. (see [here](https://github.com/rust-lang/cargo/blob/a06986cda652fb9d91ce027977d2126efa111e4f/tests/testsuite/build.rs#L6576-L6776)). That way the test can ensure that the handle is closed *before* `rustc` prints. Those tests also have some comments about the hazards of how this works.

jieyouxu replied:
Thanks for the context, I see why that test is so complex now...

jieyouxu replied:
Adjusted the test to use anonymous pipes instead which will make sure stdout is broken before the command is spawned.

GuillaumeGomez commented:
> because rustdoc is also built (both before this PR and after) with -Zon-broken-pipe=kill, and the regression test exercises the rustdoc binary and checks that it does not panic upon broken pipe. It is not clear to me if this is the desired behavior for rustdoc, hence this FYI.

I have to admit I have no idea. Being aligned with rustc seems the best course of action here.

onur-ozkan commented on src/bootstrap/src/core/build_steps/tool.rs:
```diff
@@ -209,11 +209,27 @@ pub fn prepare_tool_cargo(
-        // If the output is piped to e.g. `head -n1` we want the process to be killed,
-        // rather than having an error bubble up and cause a panic.
-        cargo.rustflag("-Zon-broken-pipe=kill");
+        cargo.env("UNTRACKED_BROKEN_PIPE_FLAG", "-Zon-broken-pipe=kill");
```
Worth to add a comment that explains why we set this environment variable instead of setting a rustflag. Above there is a lot of context, but it doesn't say anything about the workaround.

jieyouxu replied:
I did explain this above

> We use a special env var here that is not part of `RUSTFLAGS` tracked by cargo as that will cause unnecessary tool rebuilds due to tool build cache invalidation from building e.g. cargo *without* `-Zon-broken-pipe=kill` but then rustdoc *with* `-Zon-broken-pipe=kill`.

But I appreciate the wall of text makes that harder to spot, so instead I moved the workaround explanation into its own comment next to this env var. I changed the env var to be called `FORCE_ON_BROKEN_PIPE_KILL` as suggested.

jieyouxu commented on src/bootstrap/src/bin/rustc.rs:
```diff
@@ -137,6 +137,10 @@ fn main() {
 
+    if let Ok(flags) = env::var("UNTRACKED_BROKEN_PIPE_FLAG") {
+        cmd.args(flags.split_whitespace());
+    }
```
Changed to use `is_some` and renamed env var to `FORCE_ON_BROKEN_PIPE_KILL`.

jieyouxu commented:
Changes since last review:

- Renamed untracked env var from `UNTRACKED_BROKEN_PIPE_FLAG` -> `FORCE_ON_BROKEN_PIPE_KILL` as suggested, and check for env var being set then in turn set `-Zon-broken-pipe=kill` explicitly in rustc bin shim to be clearer.
- Improved clarity and locality of the comment for the untracked env var workaround.

onur-ozkan commented on src/bootstrap/src/bin/rustc.rs:
```diff
@@ -137,6 +137,15 @@ fn main() {
         cmd.args(lint_flags.split_whitespace());
     }
 
+    // Use an untracked env var (untracked by cargo) to pass `-Zon-broken-pipe=kill` to rustc when
+    // this shim is *not* used to build cargo itself, i.e. set `-Zon-broken-pipe=kill` only when
+    // building non-cargo tools.
```
We don’t need to mention `cargo` here. Rustc shim doesn’t know anything about `cargo` at this point. It’s more about where we propagate this environment variable.

jieyouxu replied:
Dropped cargo tracking here and amended the comment to only explain why the flag is conditionally passed.

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit b20436b7d3880dbaaf804093150a3931fa32e6bc with merge 983e2001e03fc9f5761f269f2cb24bfcdac566e5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b20436b7d3880dbaaf804093150a3931fa32e6bc","merge_sha":"983e2001e03fc9f5761f269f2cb24bfcdac566e5"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11219615560/job/31185980489) [(plain)](https://github.com/rust-lang-ci/rust/commit/983e2001e03fc9f5761f269f2cb24bfcdac566e5/checks/31185980489/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (49422/49921)
Updating files:  99% (49655/49921)
Updating files: 100% (49921/49921)
Updating files: 100% (49921/49921), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'983e2001e03fc9f5761f269f2cb24bfcdac566e5'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
---- [run-make] tests\run-make\broken-pipe-no-ice stdout ----

error: run-make test failed: could not build `rmake.rs` recipe
status: exit code: 1
command: "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\bin\\rustc.exe" "-o" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\broken-pipe-no-ice\\rmake.exe" "-Ldependency=C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools-bin" "-Ldependency=C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps" "-Ldependency=C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\release\\deps" "--extern" "run_make_support=C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools-bin\\librun_make_support.rlib" "--edition=2021" "C:\\a\\rust\\rust\\tests\\run-make\\broken-pipe-no-ice\\rmake.rs" "-Dunused_must_use"
--- stderr -------------------------------
error[E0433]: failed to resolve: could not find `unix` in `os`
##[error]  --> C:\a\rust\rust\tests\run-make\broken-pipe-no-ice\rmake.rs:11:14
   |
---
note: the item is gated here
  --> C:\a\rust\rust\library\std\src\os\mod.rs:19:1
   |
19 | / #[cfg(all(
20 | |     doc,
21 | |     any(
22 | |         all(target_arch = "wasm32", not(target_os = "wasi")),
23 | |         all(target_vendor = "fortanix", target_env = "sgx")
25 | | ))]
   | |___^
note: found an item that was configured out
  --> C:\a\rust\rust\library\std\src\os\mod.rs:65:9
  --> C:\a\rust\rust\library\std\src\os\mod.rs:65:9
   |
65 | pub mod unix;
   |         ^^^^
note: the item is gated here
  --> C:\a\rust\rust\library\std\src\os\mod.rs:64:1
   |
64 | #[cfg(all(not(target_os = "hermit"), any(unix, doc)))]

warning: unused import: `libc`
##[warning]  --> C:\a\rust\rust\tests\run-make\broken-pipe-no-ice\rmake.rs:14:33
   |
   |
14 | use run_make_support::{env_var, libc};
   |
   = note: `#[warn(unused_imports)]` on by default

error: aborting due to 1 previous error; 1 warning emitted
---
test result: FAILED. 280 passed; 1 failed; 86 ignored; 0 measured; 0 filtered out; finished in 96.02s

Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 1:40:38
make: *** [Makefile:106: ci-msvc-ps1] Error 1
  network time: Mon, 07 Oct 2024 18:08:56 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11219615560/job/31191596850)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11219615560/job/31191596850","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors rollup=never (subtle build system modifications and a run-make test that has platform-specific behaviors)

jieyouxu commented:
Fixed the regression test so this should now be ready for review.

@rustbot ready

ChrisDenton commented on tests/run-make/broken-pipe-no-ice/rmake.rs:
```diff
@@ -0,0 +1,73 @@
+    #[cfg(windows)]
+    {
+        match bin {
+            // On Windows, rustc has a paper that propagates the panic exit code of 101 but converts
+            // broken pipe errors into fatal errors instead of ICEs.
```
You could try changing the windows hack to a `early_fatal` error: <https://github.com/rust-lang/rust/blob/cf24c73141a77db730f4b7fda69dcd7e8b113b51/compiler/rustc_driver_impl/src/lib.rs#L1412> I think that would result in the exit code being `1`.

jieyouxu replied:
I'll do that in a follow-up, I would like to stop run-make cargo rebuilds lol

ChrisDenton replied:
Fair!

onur-ozkan commented:
@bors r+

bors commented:
:pushpin: Commit 5c0131641d2edfc21e4f627f94f0aadc6cfc6e47 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 5c0131641d2edfc21e4f627f94f0aadc6cfc6e47 -->
<!-- homu: {"type":"Approved","sha":"5c0131641d2edfc21e4f627f94f0aadc6cfc6e47","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5c0131641d2edfc21e4f627f94f0aadc6cfc6e47 with merge 18deb53874ac4701ba10ebc016cb8cde7a049b82...
<!-- homu: {"type":"BuildStarted","head_sha":"5c0131641d2edfc21e4f627f94f0aadc6cfc6e47","merge_sha":"18deb53874ac4701ba10ebc016cb8cde7a049b82"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11245328313/job/31269227948)
Approved by: onur-ozkan
Pushing 18deb53874ac4701ba10ebc016cb8cde7a049b82 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"onur-ozkan","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11245328313/job/31269227948"},"merge_sha":"18deb53874ac4701ba10ebc016cb8cde7a049b82"} -->

rust-timer commented:
Finished benchmarking commit (18deb53874ac4701ba10ebc016cb8cde7a049b82): [comparison URL](https://perf.rust-lang.org/compare.html?start=6f4ae0f34503601e54680a137c1db0b81b56cc3d&end=18deb53874ac4701ba10ebc016cb8cde7a049b82&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6f4ae0f34503601e54680a137c1db0b81b56cc3d&end=18deb53874ac4701ba10ebc016cb8cde7a049b82&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6f4ae0f34503601e54680a137c1db0b81b56cc3d&end=18deb53874ac4701ba10ebc016cb8cde7a049b82&stat=max-rss)
<details>
<summary>Results (primary -2.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.4% | [-2.4%, -2.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.4% | [-2.4%, -2.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6f4ae0f34503601e54680a137c1db0b81b56cc3d&end=18deb53874ac4701ba10ebc016cb8cde7a049b82&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=6f4ae0f34503601e54680a137c1db0b81b56cc3d&end=18deb53874ac4701ba10ebc016cb8cde7a049b82&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 774.51s -> 774.814s (0.04%)
**Artifact size**: 329.61 MiB -> 329.56 MiB (-0.02%)
